ENH: expose max_labels in ColorMap

martinfleis opened the pull request:
Closes #88

This PR introduces 2 changes to the ColorMap labels.

1. It exposes `max_labels` keyword to control how many labels should be shown. At the moment there is a hard-coded value of 10, which often causes overlaps.
2. It changes which labels are used to always have the minimum and the maximum. Current behaviour shows min, 8 in between and the second to last, not the last (max). This issue is not so apparent with 10 labels but if you use 2 you get (without this change) one in the beginning and one in the middle of the color bar which is probably not intended.

We are currently working on folium-based plotting to be included in GeoPandas and this is one of the things which would be great to fix to make a smoother user experience.

One question - how do you want to test it? I see that branca tests are essentially written as a check that it the code does not fail.

martinfleis commented:
CI failures are not related to this PR.

Conengmo commented on branca/colormap.py:
```diff
@@ -70,24 +70,27 @@ class ColorMap(MacroElement):
         The right bound of the color scale.
     caption: str
         A caption to draw with the colormap.
+    max_labels : int
```
Please also add the default value to all the docstring entries:

```suggestion
    max_labels : int, default 10
```

Conengmo commented on branca/colormap.py:
```diff
@@ -70,24 +70,27 @@ class ColorMap(MacroElement):
     caption: str
         A caption to draw with the colormap.
+    max_labels : int
+        Maximum number of legend values
```
I interpret "values" to mean something different than "ticks". Values are the values used in the colorbar, which can be more than the number of labels or ticks.

Conengmo commented on branca/utilities.py:
```diff
@@ -44,11 +44,12 @@ def legend_scaler(legend_values, max_labels=10.0):
         for i in legend_values[::spacer]:
             legend_ticks += [i]
             legend_ticks += ['']*(spacer-1)
+        legend_ticks += [legend_values[-1]]
```
I understand the desire to always show the last value. But this means we may have differing distances between labels. For example consider values `[0, 1, 2, 3]` and `max_labels=3`. Should the ticks then be `[0, 2]` (current situation, equal distance between values but missing last value) or `[0, 2, 3]` (your PR, unequal distance between values but including last value)? I don't know which is better at the moment, but in doubt I tend to stick with the status quo. What do you think?

Is there a way to have both? Include the first and last value, and have equal spacing?

Conengmo commented on branca/utilities.py:
```diff
@@ -44,11 +44,12 @@ def legend_scaler(legend_values, max_labels=10.0):
         legend_ticks = []
         for i in legend_values[::spacer]:
             legend_ticks += [i]
             legend_ticks += ['']*(spacer-1)
```
Not related to your change directly, but I don't understand why we add empty strings in between. There doesn't seem to be a need for it. If the idea is to make the length of `legend_ticks` correspond to `legend_values`, it's not even correct since we add one more tick than value.

martinfleis replied:
Don't know but commenting it out doesn't change anything apparent.

Conengmo reviewed (commented):
I'm okay with exposing this new argument, that's pretty straight forward, it doesn't yet clutter the `__init__` and setting the number of labels is a pretty important part of a colorbar.

I'll merge #93 soon, so you can adapt/extend these test cases based on your changes.

martinfleis commented:
I've added basic tests for `LinearColormap` and `StepColormap`, fixed docstrings and reverted the addition of the last value to the legend. Should be ready for another round of review/merge.

Conengmo commented on tests/test_colormap.py:
```diff
@@ -55,3 +56,28 @@ def test_step_object():
+])
+def test_max_labels_linear(max_labels, expected):
+    colorbar = cm.LinearColormap(['red'] * 10, vmin=0, vmax=9, max_labels=max_labels)
+    try:
+        colorbar.render()
+    except AssertionError:
+        assert colorbar.tick_labels == expected
```
nit: I was a bit confused by this at first. I was thinking, what does that `assert` have to do with that `AssertionError`? But I understand now that you are catching the error created because the colorbar doesn't have a Figure as parent.

In this case I would split these statements, so it's clear that the `assert` here doesn't relate to the `AssertionError` you are catching.

```suggestion
    try:
        colorbar.render()
    except AssertionError:
        pass
    assert colorbar.tick_labels == expected
```

This is a bit of a detail, so I don't mind keeping it as is.

martinfleis replied:
Sure. I'll also add a comment with a short explanation for other reading the code.

Conengmo reviewed (commented):
Changes look good 👍  You may have noticed that we are working on getting the tests running again. When they do, let's make sure this PR passes, then we can merge it. If https://github.com/python-visualization/branca/pull/95 is not enough and it'll take longer to get the tests working, we can also merge this earlier.
